src/backend/providers/playnite/PlayniteMetadataParser.cpp

```diff
@@ -0,0 +1,173 @@
+
+namespace providers {
+namespace playnite {
+
+PlayniteMetadataParser::PlayniteMetadataParser(QString log_tag, const QDir& playnite_dir)
+    : m_log_tag(std::move(log_tag))
+    , m_playnite_dir(playnite_dir)
+    , dir_filters(QDir::Files | QDir::Readable | QDir::NoDotAndDotDot)
+    , json_ext_list(QStringList() << QStringLiteral("*.json"))
+{}
+
+PlayniteComponents PlayniteMetadataParser::parse_metdata() const
+{
+    PlayniteComponents components;
+    components.platforms = parse_platform_metadata();
+    components.sources = parse_source_metadata();
+    components.companies = parse_company_metadata();
+    components.genres = parse_genre_metadata();
+    components.emulators = parse_emulator_metadata();
+    components.games = parse_game_metadata();
+    return components;
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_platform_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/platforms"));
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_source_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/sources"));
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_company_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/companies"));
+}
+
+HashMap<QString, QString> PlayniteMetadataParser::parse_genre_metadata() const
+{
+    return parse_id_name_files(QStringLiteral("library/genres"));
+}
+
+HashMap<QString, PlayniteEmulator> PlayniteMetadataParser::parse_emulator_metadata() const
+{
+    HashMap<QString, PlayniteEmulator> emulator_map;
+    QDirIterator dir_it(m_playnite_dir.filePath(QStringLiteral("library/emulators/")), json_ext_list, dir_filters);
+    while (dir_it.hasNext()) {
+        PlayniteEmulator emu;
+        const QString file_path = dir_it.next();
+        const QJsonObject json_object = get_json_object_from_file(file_path);
+        const auto emu_json = JsonObjectHelper(json_object);
+        emu.id = emu_json.get_string(QStringLiteral("Id"));
+        emu.name = emu_json.get_string(QStringLiteral("name"));
+        auto profiles = emu_json.get_json_array(QStringLiteral("Profiles"));
+        for (const auto profile : profiles) {
+            PlayniteEmulator::EmulatorProfile emu_profile;
+            auto profile_json = JsonObjectHelper(profile.toObject());
+            emu_profile.platform_ids = profile_json.get_string_list(QStringLiteral("Platforms"));
+            emu_profile.executable = profile_json.get_string(QStringLiteral("Executable"));
+            emu_profile.arguments = profile_json.get_string(QStringLiteral("Arguments"));
+            emu_profile.working_dir = profile_json.get_string(QStringLiteral("WorkingDirectory"));
+            emu_profile.id = profile_json.get_string(QStringLiteral("Id"));
+            emu_profile.name = profile_json.get_string(QStringLiteral("Name"));
+            emu.profiles.insert({ emu_profile.id, emu_profile });
+        }
+        emulator_map.insert({ emu.id, emu });
```

Comment:
You might want to make sure at least the most important fields are valid, before storing anything for further use. (Same for the other `parse_` functions.)

Reply:
I initially did all of the validation during parsing, but it became hard to read and split up validation into many different places.  That is why I let the parsing continue with empty values and then did all of the validation all in one place.

I don't think there's really much of a performance penalty for doing things this way, because by the time we have a QJsonObject we already have it backed by a QHash with all of the fields.  So each of the `get` calls is constant time.

But if you prefer I can do the validation at the same time as the parsing instead.

Reply:
Good point; I see `validate_game` for example dealing with some of it, so that's fine too. I usually prefer parse time, to avoid accidentally using them, and to keep less things in memory, but if it leads to hard to maintain code, the one place check is good too.

Reply:
I'll short circuit in the cases where the QJSonObject is empty (probably means the file doesn't exist)